readonly_bs.go

```diff
@@ -0,0 +1,207 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch shards containing the block: %w", err)
+	}
+	if len(shards) == 0 {
+		return nil, errors.New("no shards contain the requested block")
+	}
+
+	// do we have a cached blockstore for a shard containing the required block ? If yes, serve the block from that blockstore
+	for _, sk := range shards {
+		// a valid cache hit here updates the priority of the shard's blockstore in the LRU cache.
+		val, ok := ro.blockstoreCache.Get(sk)
+		if !ok {
+			continue
+		}
+
+		rbs := val.(*accessorWithBlockstore).bs
+		blk, err := rbs.Get(ctx, c)
+		if err != nil {
+			ro.blockstoreCache.Remove(sk)
+			continue
+		}
+
+		// add the block to the block cache
+		logbs.Debugw("bitswap Get: returning from block store cache", "cid", c)
+		ro.blockCache.Add(mhash.String(), blk)
+		return blk, nil
+	}
+
+	// ---- we don't have a cached blockstore for a shard that can serve the block -> let's build one.
+
+	// select a valid shard that can serve the retrieval
+	sk, err := ro.shardSelectF(c, shards)
+	if err != nil {
+		return nil, fmt.Errorf("failed to select a shard: %w", err)
+	}
+
+	// load blockstore for the selected shard and tru to serve the cid from that blockstore.
+	resch := make(chan ShardResult, 1)
+	if err := ro.d.AcquireShard(ctx, sk, resch, AcquireOpts{}); err != nil {
+		return nil, fmt.Errorf("failed to acquire shard %s: %w", sk, err)
+	}
+	var res ShardResult
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case res = <-resch:
+		if res.Error != nil {
+			return nil, fmt.Errorf("failed to acquire shard %s: %w", sk, res.Error)
+		}
+	}
+
+	sa := res.Accessor
+	bs, err := sa.Blockstore()
+	if err != nil {
+		return nil, fmt.Errorf("failed top load read only blockstore for shard %s: %w", sk, err)
+	}
+
+	blk, err := bs.Get(ctx, c)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get block: %w", err)
+	}
+
+	// update the block cache and the blockstore cache
+	ro.blockstoreCache.Add(sk, &accessorWithBlockstore{sa, bs})
+	ro.blockCache.Add(mhash.String(), blk)
+
+	logbs.Debugw("bitswap Get: returning after creating new blockstore", "cid", c)
+	return blk, nil
+}
+
+func (ro *AllShardsReadBlockstore) Has(ctx context.Context, c cid.Cid) (bool, error) {
+	logbs.Debugw("bitswap Has called", "cid", c)
+
+	// if there is a shard that can serve the retrieval for the given cid, we have the requested cid
+	// and has should return true.
+	shards, err := ro.d.ShardsContainingMultihash(ctx, c.Hash())
+	if err != nil {
+		logbs.Debugw("bitswap Has error", "cid", c, "err", err)
+		return false, fmt.Errorf("failed to fetch shards containing the multihash %w", err)
+	}
+	if len(shards) == 0 {
+		logbs.Debugw("bitswap Has: returning false no error", "cid", c)
+		return false, nil
+	}
+
+	_, err = ro.shardSelectF(c, shards)
+	if err != nil {
+		logbs.Debugw("bitswap Has error", "cid", c, "err", err)
+		return false, fmt.Errorf("failed to select a shard: %w", err)
+	}
```

Comment:
Why do we need to run the shard select function here?

Reply:
Because only blocks that has returns true for are retrievable.

Say you have configured this function to only allow deals which are available for free retrieval. Then has should return false if a given cid dosen't belong to any free deal so we're not asked to serve those blocks.

Reply:
So then should we distinguish between an error and a selection of size zero?

Reply:
Done.